junction/src/main/java/org/killbill/billing/junction/plumbing/billing/DefaultInternalBillingApi.java

```diff
@@ -220,23 +259,19 @@ private void addBillingEventsForSubscription(final ImmutableAccountData account,
                 overridenBCD = transition.getNextBillCycleDayLocal() != null ? transition.getNextBillCycleDayLocal() : overridenBCD;
                 final int bcdLocal = overridenBCD != null ?
                                      overridenBCD :
-                                     calculateBcdForTransition(catalog, bcdCache, baseSubscription, subscription, currentAccountBCD, transition, context);
+                                     calculateBcdForTransition(alignment, bcdCache, baseSubscription, subscription, currentAccountBCD, context);
 
-                if (currentAccountBCD == 0 && !updatedAccountBCD) {
-                    log.info("Setting account BCD='{}', accountId='{}'", bcdLocal, account.getId());
-                    accountApi.updateBCD(account.getExternalKey(), bcdLocal, context);
-                    updatedAccountBCD = true;
-                }
-
-                final BillingEvent event = new DefaultBillingEvent(transition, subscription, bcdLocal, account.getCurrency(), catalog);
+                final BillingEvent event = new DefaultBillingEvent(transition, subscription, bcdLocal, alignment, account.getCurrency(), catalog);
                 result.add(event);
             }
         }
     }
 
-    private int calculateBcdForTransition(final Catalog catalog, final Map<UUID, Integer> bcdCache, final SubscriptionBase baseSubscription, final SubscriptionBase subscription, final int accountBillCycleDayLocal, final EffectiveSubscriptionInternalEvent transition, final InternalTenantContext internalTenantContext)
-            throws CatalogApiException, AccountApiException, SubscriptionBaseApiException {
-        final BillingAlignment alignment = catalog.billingAlignment(getPlanPhaseSpecifierFromTransition(catalog, transition), transition.getEffectiveTransitionTime(), subscription.getStartDate());
+    private int calculateBcdForTransition(final BillingAlignment realBillingAlignment, final Map<UUID, Integer> bcdCache, final SubscriptionBase baseSubscription, final SubscriptionBase subscription, final int accountBillCycleDayLocal, final InternalTenantContext internalTenantContext) {
+        BillingAlignment alignment = realBillingAlignment;
+        if (alignment == BillingAlignment.ACCOUNT && accountBillCycleDayLocal == 0) {
```

Comment:
Not super excited about `calculateBcdForTransition` being called twice, one time faking `ACCOUNT` aligned with `SUBSCRIPTION` aligned and then one time with the real thing -- see comment above.

Reply:
This is actually unrelated to the recursive call.

Take a look at the change in `BillCycleDayCalculator.java`:

```
 public abstract class BillCycleDayCalculator {

     private static final Logger log = LoggerFactory.getLogger(BillCycleDayCalculator.class);
@@ -39,7 +41,8 @@ public abstract class BillCycleDayCalculator {
         int result = 0;
         switch (alignment) {
             case ACCOUNT:
-                result = accountBillCycleDayLocal != 0 ? accountBillCycleDayLocal : calculateOrRetrieveBcdFromSubscription(bcdCache, subscription, internalTenantContext);
+                Preconditions.checkState(accountBillCycleDayLocal != 0, "Account BCD should be set at this point");
+                result = accountBillCycleDayLocal;
                 break;
             case BUNDLE:
                 result = calculateOrRetrieveBcdFromSubscription(bcdCache, baseSubscription, internalTenantContext);
```

Previously, when the BCD wouldn't be set, `BillCycleDayCalculator.calculateBcdForAlignment` would call `calculateOrRetrieveBcdFromSubscription` for `ACCOUNT` alignments (this is how we compute the account BCD). I've simply moved that logic outside, so that it's explicit (and that codepath should only be exercised by junction now, while today it affects anyone calling `calculateBcdForAlignment` -- hence #865).